### PR TITLE
fix(reprocessing): Apply fresh grouping config [ISSUE-1415]

### PR DESCRIPTION
**Pull request by jjbayer**
Reprocessing is a feature for native platforms intended to re-apply debug files, as stated in our [docs](https://docs.sentry.io/product/issues/reprocessing/):

> Reprocessing allows you to apply debug information files to error events that have already occurred.

The behavior of reprocessing with regards to grouping configuration has been inconsistent: While fingerprinting rules were always taken fresh from the project, stack trace rules and the grouping strategy were taken from the event:

https://github.com/getsentry/sentry/blob/c69ac2e5d8962499659c3b22608911927782d366/src/sentry/event_manager.py#L1716

https://github.com/getsentry/sentry/blob/c69ac2e5d8962499659c3b22608911927782d366/src/sentry/event_manager.py#L392-L394

This PR aligns this behavior by always loading the grouping config from the project when handling a reprocessed event.

### CODE REVIEW

File: src/sentry/event_manager.py [hunk 1]
```diff
@@ -845,12 +854,12 @@ def _nodestore_save_many(jobs):
...
             event = job["event"]
-            data = event_processing_store.get(
+            unprocessed = event_processing_store.get(
                 cache_key_for_event({"project": event.project_id, "event_id": event.event_id}),
                 unprocessed=True,
             )
-            if data is not None:
-                subkeys["unprocessed"] = data
+            if unprocessed is not None:
+                subkeys["unprocessed"] = unprocessed
```
Comment (jjbayer) [inline src/sentry/event_manager.py:862]: This is an unrelated change.

File: tests/sentry/tasks/test_reprocessing2.py [hunk 2]
```diff
@@ -434,3 +438,154 @@ def test_nodestore_missing(
...
+    new_rules = FingerprintingRules.from_config_string(
+        """
+    message:"hello world 1" -> hw1 title="HW1"
+    """
+    )
+
+    with mock.patch(
+        "sentry.event_manager.get_fingerprinting_config_for_project", return_value=new_rules
+    ):
+        # TODO: test stack trace rules as well
```
Comment (untitaker) [inline tests/sentry/tasks/test_reprocessing2.py:485]: probably can be removed

**Review by untitaker — APPROVED** [review]

File: tests/sentry/tasks/test_reprocessing2.py (hunk 2 above)
Comment (jjbayer) [reply tests/sentry/tasks/test_reprocessing2.py:485]: Thanks!

File: src/sentry/event_manager.py [hunk 3]
```diff
@@ -389,9 +389,18 @@ def save(
         with metrics.timer("event_manager.load_grouping_config"):
             # At this point we want to normalize the in_app values in case the
             # clients did not set this appropriately so far.
-            grouping_config = get_grouping_config_dict_for_event_data(
-                job["event"].data.data, project
-            )
+            if is_reprocessed:
+                # The customer might have changed grouping enhancements since
+                # the event was ingested -> make sure we get the fresh one for reprocessing.
+                grouping_config = get_grouping_config_dict_for_project(project)
```
Comment (iker-barriocanal) [inline src/sentry/event_manager.py:395]: Question: is there a chance of getting a cached version (and thus, not a fresh one) of the config?

Comment (jjbayer) [reply src/sentry/event_manager.py:395]: The config is fetched from project options, which does have a cache, but that one is updated when a new value is written, so it should always be fresh https://github.com/getsentry/sentry/blob/6b0b3d7657ff64eb0877d2f69e4031d5c139472b/src/sentry/models/options/project_option.py#L50
